### DISCUSSION TITLE
Sending with offline_access scope to auth server when user is not in that role causes error 400

### DISCUSSION THREAD

**Discussion by ArthurTimofey** [keycloak/keycloak#24631] [Q&A] [answered]
Hi,

It seems like I am maybe misunderstanding how to work with the scope of offline_access for my realm. Because we are running into an issue when we have the following happening:

1. User logs into account through [react-oidc-context](https://github.com/authts/react-oidc-context) adapter
2. We provide following scopes: `['openid', 'offline_access']`
3. Now, IF user has role of `offline_access` added in keycloak server then it works fine and authentication is completed
4. BUT, If the user does not have the role `offline_access` yet we still have that scope in our scopes array for our adapter we run into a 400 error.

Any ideas as to what we are doing wrong?

We are using KeyCloak v22

Thank you!

**Comment by danielFesenmeyer** [comment, marked as answer]
As far as I understand, the described behavior makes sense.
If the user does not have the role `offline_access`, the scope `offline_access` is not supported for any request of that user -> `Bad Request`. Maybe `Forbidden` would also make sense as response.

**Reply by ArthurTimofey** [reply]
@danielFesenmeyer so then what is the correct way to control who has offline_access and who does not? Like if I log in and I am a user that has a valid offline_access role then I must relog again on my app in order to first test that the offline_access role is part of the profile? Or am I completely fumbling my understanding here?

**Reply by danielFesenmeyer** [reply]
What I think makes sense: Define `offline_access` as a realm default role (I think it is already configured as such as initial realm setting). Then it's automatically available for all users.

I would define per client (via scope), whether offline access should be possible or not.

But maybe @sschu has some additional hint?

**Reply by sschu** [reply]
It is as @danielFesenmeyer says, offline_access is part of the realm default role so every user has it and clients should be able to request offline_access scope. You could then go ahead and remove the role from users if you want to deny them offline tokens (or from the client scopes if you don't want certain clients to get offline tokens), but with default settings it should just work for any user/client.

**Reply by ArthurTimofey** [reply]
Thank you @danielFesenmeyer and @sschu that seems to make sense. In our case we would need to have two different clients to do this properly. One for general user that does not require offline_access tokens and one for special users that need to be online 24/7. Thank you again for the clarification :)
